plutus-core/plutus-core/src/Crypto.hs

```diff
@@ -17,3 +34,58 @@ verifySignature pubKey msg sig =
             <$> publicKey pubKey
             <*> pure msg
             <*> signature sig
+
+-- | Verify an ECDSA signature made using the SECP256k1 curve.
+--
+-- = Note
+--
+-- This takes a message /hash/, rather than a general blob of bytes; thus, it is
+-- limited in length.
+verifyEcdsaSecp256k1Signature
+  :: BS.ByteString -- ^ Public key (64 bytes)
+  -> BS.ByteString -- ^ Message hash (32 bytes)
+  -> BS.ByteString -- ^ Signature (64 bytes)
+  -> Emitter (EvaluationResult Bool)
+verifyEcdsaSecp256k1Signature pk msg sig =
+  case DSIGN.rawDeserialiseVerKeyDSIGN @EcdsaSecp256k1DSIGN pk of
+    Nothing -> failWithMessage loc "Invalid verification key."
+    Just pk' -> case DSIGN.rawDeserialiseSigDSIGN @EcdsaSecp256k1DSIGN sig of
```

Comment:
You use here the definitions that we provide, but evidently what we provide is not suitable for the task. It would be reasonable to expect `failWithMessage loc str *> b` to fail with `EvaluationFailure` without evaluating `b`, but it will evaluate `b`. The problem is with our abstractions, not your code, so my recommendation would be to add a short note saying that existing abstractions are not sufficient and the code needs to be improved and then someone (probably from the Plutus team) would go and actually fix the code with a monad transformer or something. I don't think that should block this PR, since the code looks perfectly fine, it's only the abstractions that are off.

Reply:
This is slightly orthogonal to my work though - I only used `failWithMessage` as a refactor for a common 'terminating procedure', never expecting it to be 'sequenced' with anything else. I'm happy to leave such a note, but where would it go? `failWithMessage` isn't the only way to exhibit this issue, and if I put it here, future readers aren't likely to find it.

Reply:
> This is slightly orthogonal to my work though

Absolutely.

> never expecting it to be 'sequenced' with anything else.

Yes, I understand that, but a bad abstraction is a bad abstraction and we should fix it. You're doing the best that you can here with the tools you're given, I do not suggest you should change anything.

> I'm happy to leave such a note, but where would it go? `failWithMessage` isn't the only way to exhibit this issue, and if I put it here, future readers aren't likely to find it.

It's not for future readers, it's for us to keep us annoyed, so that we're motivated enough to fix the issue. Any place would work, so `failWithMessage` is good enough.